Use Markable instead of std::optional when storing UUIDs in structs

achristensen07 opened the pull request:
#### 4c52b0658f9cb5c07e054f4ae1ffe7788da7be3e
<pre>
Use Markable instead of std::optional when storing UUIDs in structs
<a href="https://bugs.webkit.org/show_bug.cgi?id=251485">https://bugs.webkit.org/show_bug.cgi?id=251485</a>
rdar://104902597

Reviewed by Yusuke Suzuki.

This removes unnecessary and unused padding bytes, decreasing memory use.

* Source/WTF/wtf/UUID.h:
(WTF::UUID::MarkableTraits::isEmptyValue):
(WTF::UUID::MarkableTraits::emptyValue):
* Source/WebCore/Modules/push-api/PushSubscriptionIdentifier.h:
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/Platform/IPC/DaemonCoders.h:
* Source/WebKit/Shared/WebPushDaemonConnectionConfiguration.h:
* Source/WebKit/Shared/WebPushDaemonConnectionConfiguration.serialization.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::identifier const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::identifier const):
* Source/WebKit/webpushd/PushClientConnection.h:
(WebPushD::ClientConnection::dataStoreIdentifier const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: <a href="https://commits.webkit.org/259665@main">https://commits.webkit.org/259665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2b2f9750d8453dd307eabcd69929b751368f0f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105627 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14717 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38501 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114855 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109529 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16123 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5916 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111381 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/97896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26879 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/95362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7986 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93526 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5770 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/8488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30405 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14104 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47780 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/102238 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6675 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/10039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25491 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
```diff
@@ -117,7 +117,7 @@ std::optional<NetworkSessionCreationParameters> NetworkSessionCreationParameters
         return std::nullopt;
     
-    std::optional<std::optional<UUID>> dataStoreIdentifier;
+    std::optional<Markable<UUID>> dataStoreIdentifier;
```
We can just decode it into `std::optional`, and then we can assign it to markable.

achristensen07 replied:
This is more correct.  I really ought to just make it NetworkSessionCreationParameters.serialization.in which would generate this, but that should probably be a separate PR.

Constellation commented on Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
```diff
@@ -50,7 +51,7 @@ class WebExtensionControllerConfiguration : public API::ObjectImpl<API::Object::
     explicit WebExtensionControllerConfiguration(const UUID&);
 
-    std::optional<UUID> identifier() const { return m_identifier; }
+    Markable<UUID> identifier() const { return m_identifier; }
```
You can continue returning std::optional if it is better since Markable <-> std::optional conversion can happen implicitly. So, I would recommend using std::optional usually, and using Markable only when we store it to the member.

Constellation commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
```diff
@@ -59,7 +60,7 @@ class WebsiteDataStoreConfiguration : public API::ObjectImpl<API::Object::Type::
 
     bool isPersistent() const { return m_isPersistent == IsPersistent::Yes; }
-    std::optional<UUID> identifier() const { return m_identifier; }
+    Markable<UUID> identifier() const { return m_identifier; }
```
Ditto.

Constellation commented on Source/WebKit/webpushd/PushClientConnection.h:
```diff
@@ -66,7 +66,7 @@ class ClientConnection : public RefCounted<ClientConnection>, public CanMakeWeak
 
     const String& pushPartitionString() const { return m_pushPartitionString; }
-    std::optional<UUID> dataStoreIdentifier() const { return m_dataStoreIdentifier; }
+    Markable<UUID> dataStoreIdentifier() const { return m_dataStoreIdentifier; }
```
Ditto.

Constellation commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
```diff
@@ -529,7 +529,7 @@ async function disableShowNotifications()
     }
 
-    const std::optional<UUID>& dataStoreIdentifier() { return m_dataStoreIdentifier; }
+    const Markable<UUID>& dataStoreIdentifier() { return m_dataStoreIdentifier; }
```
Ditto.

achristensen07 committed:
Use Markable instead of std::optional when storing UUIDs in structs
https://bugs.webkit.org/show_bug.cgi?id=251485
rdar://104902597

Reviewed by Yusuke Suzuki.

This removes unnecessary and unused padding bytes, decreasing memory use.

* Source/WTF/wtf/UUID.h:
(WTF::UUID::MarkableTraits::isEmptyValue):
(WTF::UUID::MarkableTraits::emptyValue):
* Source/WebCore/Modules/push-api/PushSubscriptionIdentifier.h:
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/Platform/IPC/DaemonCoders.h:
* Source/WebKit/Shared/WebPushDaemonConnectionConfiguration.h:
* Source/WebKit/Shared/WebPushDaemonConnectionConfiguration.serialization.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::identifier const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::identifier const):
* Source/WebKit/webpushd/PushClientConnection.h:
(WebPushD::ClientConnection::dataStoreIdentifier const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: https://commits.webkit.org/259665@main
